langfuse-core/src/index.ts

```diff
@@ -674,14 +687,207 @@ abstract class LangfuseCoreStateless {
     }
   }
 
+  protected async processMediaInEvent(type: LangfuseObject, body: EventBody): Promise<void> {
+    if (!body) return;
+
+    const traceId = "traceId" in body ? body.traceId : type.includes("trace") ? body.id : undefined;
+
+    if (!traceId) {
+      throw new Error("traceId is required for media upload");
+    }
+
+    const observationId = (type.includes("generation") || type.includes("span")) && body.id ? body.id : undefined;
+
+    await Promise.all(
+      (["input", "output", "metadata"] as const).map(async (field) => {
+        if (body[field as keyof EventBody]) {
+          body[field as keyof EventBody] =
+            (await this.findAndProcessMedia({
+              data: body[field as keyof EventBody],
+              traceId,
+              observationId,
+              field,
+            }).catch((e) => {
+              this._events.emit("error", `Error processing multimodal event: ${e}`);
+            })) ?? body[field as keyof EventBody];
+        }
+      })
+    );
+  }
+
+  protected async findAndProcessMedia({
+    data,
+    traceId,
+    observationId,
+    field,
+  }: {
+    data: any;
+    traceId: string;
+    observationId?: string;
+    field: "input" | "output" | "metadata";
+  }): Promise<any> {
+    const seenObjects = new WeakMap();
+    const maxLevels = 10;
+
+    const processRecursively = async (data: any, level: number): Promise<any> => {
+      if (typeof data === "string" && data.startsWith("data:")) {
+        const media = new LangfuseMedia({ base64DataUri: data });
+        await this.processMediaItem({ media, traceId, observationId, field });
+
+        return media;
+      }
+      if (typeof data !== "object" || data === null) {
+        return data;
+      }
+
+      // Use WeakMap to detect cycles
+      if (seenObjects.has(data) || level > maxLevels) {
+        return data;
+      }
```

Comment:
should this not be at the top of the function?

Reply:
weakmap does not support primitives, so need to handle primitives first (we don't recurse them anyways) prior to handling objects